### ISSUE TITLE
IP tunnels fail to correctly persist across zone reboots

### ISSUE THREAD

**Issue by brianewell** [joyent/smartos-live#737]
Any IP tunnels configured with `dladm` and `ipadm` in a zone will become inconsistent and unusable without manual intervention after the zone is rebooted.

```
gz# zlogin 41a151f1-7776-6d9a-d7f9-f424272a7a2e
zone# dladm create-iptun -T ipv4 -a local=192.168.0.179,remote=192.168.0.55 v4_test0
zone# dladm show-iptun
LINK            TYPE  FLAGS  LOCAL               REMOTE
v4_test0        ipv4  --     192.168.0.179       192.168.0.55
zone# ipadm create-addr -T addrconf v4_test0/v6
zone# ipadm show-addr
ADDROBJ           TYPE     STATE        ADDR
lo0/v4            static   ok           127.0.0.1/8
net0/?            dhcp     ok           192.168.0.179/24
lo0/v6            static   ok           ::1/128
v4_test0/v6       addrconf ok           fe80::c0a8:b3->fe80::c0a8:37
zone# logout

[Connection to zone '41a151f1-7776-6d9a-d7f9-f424272a7a2e' pts/2 closed]
gz# vmadm reboot 41a151f1-7776-6d9a-d7f9-f424272a7a2e
Successfully completed reboot for VM 41a151f1-7776-6d9a-d7f9-f424272a7a2e
gz# zlogin 41a151f1-7776-6d9a-d7f9-f424272a7a2e
zone# dladm show-iptun
zone# ipadm show-addr
ADDROBJ           TYPE     STATE        ADDR
lo0/v4            static   ok           127.0.0.1/8
net0/?            dhcp     ok           192.168.0.179/24
lo0/v6            static   ok           ::1/128
v4_test0/v6       addrconf disabled     ::
```

This behavior appears to have started with platform image 20170202T040152Z.  The behavior on image 20170119T014200Z is as expected, with the configuration coming right back up after a reboot.

**Comment by jclulow** [comment]
This looks like something @danmcd might know about!

**Comment by danmcd** [comment]
SmartOS completely changes the network/physical and other SMF methods that normally use dladm and ipadm.  dladm & ipadm don't work for ANY type of interface in a zone, not just IP tunnels.  Even the old-fashioned /etc/hostname.* files won't work.

As a workaround in my own SmartOS zone for IP tunnels, I had to put in an /etc/rc3.d script.

The proper solution would be to inspect the design decisions made for IP interface configuration, and see how the existing tools' state (/etc/dladm/* and /etc/ipadm/*) can be queried for cases of IP tunnelling.

**Comment by brian-ewell** [comment]
So then it's an accident that it works in 20170119T014200Z?

**Comment by danmcd** [comment]
Interesting.  I've only recently (June) arrived at full-time SmartOS usage.  But given things changed after the build you mentioned, I'm curious about that myself.  I'll have to dig into the diffs and see what all happened.  In general, SmartOS prefers zone network configuration by zonecfg(1M)/vmadm(1M) vs. in-zone configuration like other illumos distros.

**Comment by danmcd** [comment]
You don't happen to know if this worked on images prior to 20170119... ?

**Comment by brianewell** [comment]
It had been working for years (since some time in 2015) before then neither `dladm` or `ipadm` would accept configurations except by using the temporary flag (`-t`).

A bunch of dladm related commits from @rzezeski had been included in the 20170202T040152Z update (joyent@d7a890514240ff00bf92194bd9186dc4fb68a367, joyent@0e21e957f7ee4ef75f79e71f2fd6f09da211893d, joyent@33df115038809a05a8856b165fabdb823350567e, joyent@aefd5a8c21e439a1955ccb6b19739e9bf7e8cec3), it might be a good idea to include him in this issue as well.

**Comment by brianewell** [comment]
I suppose this is as good a time as any to learn Illumos internals.

**Comment by danmcd** [comment]
Specifically the SmartOS modifications.  I'd start with ipmgmtd, dlmgtmd, and zoneadmd.

**Comment by rzezeski** [comment]
@brianewell Thank you for the report. This turned out to be a subtle regression introduced in joyent/illumos-joyent@33df115. I'll have a full report tomorrow and will try to get a fix in this week.

**Comment by rzezeski** [comment]
A full write up can be found at [OS-6274](https://smartos.org/bugview/OS-6274).

**Comment by sjorge** [comment]
@rzezeski just wanted to say, awesome write up in OS-6274!

**Comment by brianewell** [comment]
@rzezeski Wow, that write up was incredible.  My debugging mirrored yours right up to about 5% into the report, where you used `mdb`.  It's going to take some time to make sense of the next 95%.

I have some basic knowledge of mdb and dtrace, but reading that report was like listening to a symphony.

**Comment by rzezeski** [comment]
@brianewell This is fixed as of joyent/illumos-joyent@c08af995b7a4018333aa45db89575979b416b60a and should be included in the next release. Thanks again for taking the time to file a report and my apologies for any inconvenience this has caused.
